monai/metrics/seg_metric_utils.py

```diff
@@ -0,0 +1,220 @@
+
+    Raises:
+        NotImplementedError: When input an activation name that is not implemented.
+    """
+    input_ndim = input_data.ndimension()
+    if activation == "softmax":
+        if input_ndim == 1:
+            warnings.warn("input_data has only one channel, softmax ignored.")
+        else:
+            input_data = input_data.float().softmax(dim=1)
+    elif activation == "sigmoid":
+        input_data = input_data.float().sigmoid()
+    elif callable(activation):
+        input_data = activation(input_data)
+    else:
+        raise NotImplementedError("activation can only be sigmoid, softmax or a callable function.")
+    return input_data
+
+
+def do_binarization(
+    input_data: torch.Tensor,
+    bin_mode: str = "threshold",
+    bin_threshold: Union[float, Sequence[float]] = 0.5,
+) -> torch.Tensor:
+    """
+    Args:
+        input_data: the input that to be binarized, in the shape [B] or [BN] or [BNHW] or [BNHWD].
+        bin_mode: can be ``"threshold"`` or ``"mutually_exclusive"``, or a callable function.
+            - ``"threshold"``, a single threshold or a sequence of thresholds should be set.
+            - ``"mutually_exclusive"``, `input_data` will be converted by a combination of
+            argmax and to_onehot.
+        bin_threshold: the threshold to binarize the input data, can be a single value or a sequence of
+            values that each one of the value represents a threshold for a class.
+
+    Raises:
+        AssertionError: when `bin_threshold` is a sequence and the input has the shape [B].
+        AssertionError: when `bin_threshold` is a sequence but the length != the number of classes.
+        AssertionError: when `bin_mode` is ``"mutually_exclusive"`` the input has the shape [B].
+        AssertionError: when `bin_mode` is ``"mutually_exclusive"`` the input has the shape [B, 1].
+    """
+    input_ndim = input_data.ndimension()
+    if bin_mode == "threshold":
+        if isinstance(bin_threshold, Sequence):
+            assert input_ndim > 1, "a sequence of thresholds are used for multi-class tasks."
+            error_hint = "the length of the sequence should be the same as the number of classes."
+            n_classes = input_data.shape[1]
+            assert n_classes == len(bin_threshold), "{}".format(error_hint)
+            for cls_num in range(n_classes):
+                input_data[:, cls_num] = (input_data[:, cls_num] >= bin_threshold[cls_num]).float()
+        else:
+            input_data = (input_data >= bin_threshold).float()
+    elif bin_mode == "mutually_exclusive":
+        assert input_ndim > 1, "mutually_exclusive is used for multi-class tasks."
+        n_classes = input_data.shape[1]
+        assert n_classes > 1, "mutually_exclusive is used for multi-class tasks."
+        input_data = torch.argmax(input_data, dim=1, keepdim=True)
+        input_data = one_hot(input_data, num_classes=n_classes)
+    return input_data
+
+
+def preprocess_input(
```

Comment:
I think we can remove all the pre-processing logic for metrics, and ask users to use `post_transforms` before metrics computation. Otherwise, they have duplicated parts and confusing.
@wyli @ericspod, what do you think? If you agree, we can also update examples and tutorials to use `post_transforms`.  Actually, if we changed to use `post_transforms` in examples, the examples are still compatible with MONAI v0.3.0.

Thanks.

Reply:
Yes, making things more modular makes sense, putting general purpose functionality into post transforms and expecting users to apply them as appropriate is the way to go with this.